File: hashTable.py
class HashTable:
    def __init__(self, size=10):
        self.table_size = size  
        self.table = [[] for _ in range(self.table_size)]
        self.count = 0

    def hash(self, key):
        hash_value = 0
        for char in key:
            hash_value = (hash_value * 31 + ord(char)) % self.table_size
        return hash_value

    def add(self , ch):
        key = ch.split()[-1]
        index = self.hash(key)
        if key in self.table[index]:
            print(f"already added")
        elif len(ch.split()) >2:
            print("Invalid input")    
        else:
            self.table[index].append(key)
            self.count += 1 
            print("added ")

    def remove(self , ch):
        ch.split()[-1]
        index = self.hash(ch.split()[-1])
        try:    
            if len(ch.split())<2:
                print("Invalid input")
            else:    
                self.table[index].remove(ch.split()[-1])
                self.count -= 1
                print("removed ")
        except ValueError:
            print(f"{ ch.split()[-1]} not found")

    def get_size(self):  
        return self.count

   
    def contains(self, ch):
        key = ch.split()[-1]
        index = self.hash(key)
        if key in self.table[index]:
            print(f"{key} found")
            return True
        else:
            print(f"{key} not found")
            return False

    def display(self):
        for i in range(self.table_size):
            print(f"{i}.- {self.table[i]}")


hash_table = HashTable()

while True:
    ch = input(">").split()
    if ch[0]=="add":
        hash_table.add(ch[-1])
    elif ch[0]=="remove":
        hash_table.remove(ch[-1])
    elif ch[0]=="contains":
        hash_table.contains(ch[-1])
    elif ch[0]=="size":
        print(f"{hash_table.get_size()}")
    elif ch[0]=="display":
        hash_table.display()
    elif ch[0] == "exit":
        break
    else:
        print("Invalid operation")

 

 

File: HashTable.java
package org.example;
import java.util.*;

public class HashTable {
    private int tableSize;
    private ArrayList<String>[] table;
    private int count;

    public HashTable(int size) {
        this.tableSize = size;
        this.table = new ArrayList[tableSize];
        for (int i = 0; i < tableSize; i++) {
            table[i] = new ArrayList<>();
        }
        this.count = 0;
    }

    public HashTable() {
        this(10);
    }

    private int hash(String key) {
        int hashValue = 0;
        for (char ch : key.toCharArray()) {
            hashValue = (hashValue * 31 + ch) % tableSize;
        }
        return hashValue;
    }

    public void add(String ch) {
        String[] parts = ch.split(" ");
        if (parts.length != 2) {
            System.out.println("Invalid input");
            return;
        }
        String key = parts[1];
        int index = hash(key);
        if (table[index].contains(key)) {
            System.out.println("Already added");
        } else {
            table[index].add(key);
            count++;
            System.out.println("Added");
        }
    }

    public void remove(String ch) {
        String[] parts = ch.split(" ");
        if (parts.length != 2) {
            System.out.println("Invalid input");
            return;
        }
        String key = parts[1];
        int index = hash(key);
        if (table[index].remove(key)) {
            count--;
            System.out.println("Removed");
        } else {
            System.out.println(key + " not found");
        }
    }

    public int getSize() {
        return count;
    }

    public boolean contains(String ch) {
        String[] parts = ch.split(" ");
        if (parts.length != 2) {
            System.out.println("Invalid input");
            return false;
        }
        String key = parts[1];
        int index = hash(key);
        if (table[index].contains(key)) {
            System.out.println(key + " found");
            return true;
        } else {
            System.out.println(key + " not found");
            return false;
        }
    }

    public void display() {
        for (int i = 0; i < tableSize; i++) {
            System.out.println(i + ".- " + table[i]);
        }
    }

    public static void main(String[] args) {
        HashTable hashTable = new HashTable();
        Scanner scanner = new Scanner(System.in);
        while (true) {
            System.out.print(">");
            String input = scanner.nextLine();
            String command = input.split(" ")[0];
            if (command.equals("add")) {
                hashTable.add(input);
            } else if (command.equals("remove")) {
                hashTable.remove(input);
            } else if (command.equals("size")) {
                System.out.println(hashTable.getSize());
            } else if (command.equals("contains")) {
                hashTable.contains(input);
            } else if (command.equals("display")) {
                hashTable.display();
            } else if (command.equals("exit")) {
                break;
            } else {
                System.out.println("Invalid command");
            }
        }
    }
}
